src/client/dfs/dfs.c

```diff
@@ -3900,6 +3945,131 @@ dfs_chmod(dfs_t *dfs, dfs_obj_t *parent, const char *name, mode_t mode)
 	return rc;
 }
 
+int
+dfs_chown(dfs_t *dfs, dfs_obj_t *parent, const char *name, uid_t uid, gid_t gid)
+{
+	daos_handle_t		oh;
+	daos_handle_t		th = DAOS_TX_NONE;
+	bool			exists;
+	struct dfs_entry	entry = {0};
+	daos_key_t		dkey;
+	d_sg_list_t		sgl;
+	d_iov_t			sg_iovs[2];
+	daos_iod_t		iod;
+	daos_recx_t		recx[2];
+	size_t			len;
+	dfs_obj_t		*sym;
+	const char		*entry_name;
+	int			i;
+	int			rc;
+
+	if (dfs == NULL || !dfs->mounted)
+		return EINVAL;
+	if (dfs->layout_v <= 2) {
+		D_ERROR("DFS Container has a layout version that does not support dfs_chown()\n");
+		return ENOTSUP;
+	}
+	if (dfs->amode != O_RDWR)
+		return EPERM;
+	if (parent == NULL)
+		parent = &dfs->root;
+	else if (!S_ISDIR(parent->mode))
+		return ENOTDIR;
+	if (name == NULL) {
+		if (strcmp(parent->name, "/") != 0) {
+			D_ERROR("Invalid path %s and entry name is NULL)\n",
+				parent->name);
+			return EINVAL;
+		}
+		name = parent->name;
+		len = strlen(name);
+		oh = dfs->super_oh;
+	} else {
+		rc = check_name(name, &len);
+		if (rc)
+			return rc;
+		oh = parent->oh;
+	}
+
+	/* Check if parent has the entry */
+	rc = fetch_entry(dfs->layout_v, oh, DAOS_TX_NONE, name, len, true, &exists, &entry,
+			 0, NULL, NULL, NULL);
+	if (rc)
+		D_GOTO(out, rc);
+
+	if (!exists)
+		D_GOTO(out, rc = ENOENT);
+
+	if (uid == -1 && gid == -1)
+		D_GOTO(out, rc = 0);
```

Comment:
Should this check be moved up to before fetch? But I guess if the entry didn't exist we shouldn't `return 0`. Hmm...

Reply:
i think if the entry does not exist, we want to fail.